src/charm.py

```diff
@@ -0,0 +1,576 @@
+        self._bessd_container_name = self._bessd_service_name = "bessd"
+        self._routectl_container_name = self._routectl_service_name = "routectl"
+        self._web_container_name = self._web_service_name = "web"
+        self._pfcp_agent_container_name = self._pfcp_agent_service_name = "pfcp-agent"
+        self._bessd_container = self.unit.get_container(self._bessd_container_name)
+        self._routectl_container = self.unit.get_container(self._routectl_container_name)
+        self._web_container = self.unit.get_container(self._web_container_name)
+        self._pfcp_agent_container = self.unit.get_container(self._pfcp_agent_container_name)
+        self._metrics_endpoint = MetricsEndpointProvider(
+            self,
+            jobs=[
+                {
+                    "static_configs": [{"targets": [f"*:{PROMETHEUS_PORT}"]}],
+                }
+            ],
+        )
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="pfcp", port=8805, protocol="UDP"),
+                ServicePort(name="bess-web", port=BESS_WEB_PORT),
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+            ],
+        )
+        network_attachment_definition_spec = {
+            "config": json.dumps(
+                {
+                    "cniVersion": "0.3.1",
+                    "type": "macvlan",
+                    "ipam": {"type": "static"},
+                    "capabilities": {"mac": True},
+                }
+            )
+        }
+        self._kubernetes_multus = KubernetesMultusCharmLib(
+            charm=self,
+            containers_requiring_net_admin_capability=[self._bessd_container_name],
+            network_attachment_definitions=[
+                NetworkAttachmentDefinition(
+                    metadata=ObjectMeta(name=ACCESS_NETWORK_ATTACHMENT_DEFINITION_NAME),
+                    spec=network_attachment_definition_spec,
+                ),
+                NetworkAttachmentDefinition(
+                    metadata=ObjectMeta(name=CORE_NETWORK_ATTACHMENT_DEFINITION_NAME),
+                    spec=network_attachment_definition_spec,
+                ),
+            ],
+            network_annotations_func=self._network_annotations_from_config,
+        )
+        self.framework.observe(self.on.bessd_pebble_ready, self._configure)
+        self.framework.observe(self.on.routectl_pebble_ready, self._configure)
+        self.framework.observe(self.on.web_pebble_ready, self._configure)
+        self.framework.observe(self.on.pfcp_agent_pebble_ready, self._configure)
+        self.framework.observe(self.on.config_changed, self._configure)
+        self.framework.observe(self.on.update_status, self._update_status)
+
+    def _update_status(self, event: UpdateStatusEvent):
+        """Sets unit status based on current state."""
+        if self.unit.status != ActiveStatus():
+            self._configure(event)
```

Comment:
This is interesting, was this required to fix the issue you had before?

It seems potentially wasteful, and with all the potential `defer`, couldn't this pile up a bunch of deferred events?

Reply:
No, this wasn't needed and you're right, we may want to remove it. It was more of a "protection". I'll remove it and we can add something along those lines if we really need it in the future.